File: ios/Crypto.m
#import "CryptoBindings.h"
#import "CryptoBridge.h"
#import "BCHelpers.h"
#include <sys/sysctl.h>
#import <React/RCTLog.h>

#import "Crypto.h"

@implementation Crypto

+ (NSString *)decryptText:(NSString *)text usingKey:(NSString *)keyHex andSalt:(NSString *)salt andIV:(NSString *)ivHex andHMACKey:(NSString *)hmacKeyHex andHMAC:(NSString *)hmacHex {
    const char* decryptedText = decrypt_cbc(
        [text UTF8String],
        [keyHex UTF8String],
        [ivHex UTF8String],
        [salt UTF8String],
        [hmacKeyHex UTF8String],
        [hmacHex UTF8String]
    );
    if (decryptedText == NULL) {
        return @"Error:Failed decrypting content";
    }
    NSString *output = [NSString stringWithUTF8String:decryptedText];
    dealloc_memory(decryptedText);
    return output;
}

+ (NSString *)encryptText:(NSString *)text usingKey:(NSString *)keyHex andSalt:(NSString *)salt andIV:(NSString *)ivHex andHMACKey:(NSString *)hmacKeyHex {
    const char* encryptedText = encrypt_cbc(
        [text UTF8String],
        [keyHex UTF8String],
        [salt UTF8String],
        [ivHex UTF8String],
        [hmacKeyHex UTF8String]
    );
    NSString *output = [NSString stringWithUTF8String:encryptedText];
    dealloc_memory(encryptedText);
    return output;
}

+ (NSString *)generateIVWithLength:(int)length {
    const char* iv = generate_random_bytes(length);
    NSString *output = [NSString stringWithUTF8String:iv];
    dealloc_memory(iv);
    return output;
}

+ (NSString *)generateSaltWithLength:(int)length {
    const char* salt = generate_salt(length);
    NSString *output = [NSString stringWithUTF8String:salt];
    dealloc_memory(salt);
    return output;
}

+ (NSString *)generateUUIDs:(int)count {
    const char* uuidList = generate_uuid_list(count);
    NSString *output = [NSString stringWithUTF8String:uuidList];
    dealloc_memory(uuidList);
    return output;
}

+ (NSString *)pbkdf2UsingPassword:(NSString *)password andSalt:(NSString *)salt andIterations:(int)iterations andBits:(int)bits {
    const char* keyDerivationInfo = pbkdf2_derive(
        [password UTF8String],
        [salt UTF8String],
        iterations,
        bits
    );
    NSString *output = [NSString stringWithUTF8String:keyDerivationInfo];
    dealloc_memory(keyDerivationInfo);
    return output;
}

@end

File: android/app/src/main/java/com/buttercup/Crypto.java
package com.buttercup;

import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;

import java.net.URLDecoder;
import java.lang.String;

public class Crypto extends ReactContextBaseJavaModule {
    private static native String generateUUIDList(int count);
    private static native String deriveKeyFromPassword(String password, String salt, int rounds, int bits);
    private static native String generateSalt(int length);
    private static native String generateRandomBytes(int length);
    private static native String encryptCBC(
        String encodedText,
        String keyHex,
        String salt,
        String ivHex,
        String hmacKeyHex
    );
    private static native String decryptCBC(
        String encryptedText,
        String keyHex,
        String ivHex,
        String salt,
        String hmacKeyHex,
        String hmacHex
    );

    static {
        System.loadLibrary("crypto");
    }

    public Crypto(ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @ReactMethod
    public void pbkdf2(String password, String salt, int rounds, int bits, Promise promise) {
        String derivationHex = Crypto.deriveKeyFromPassword(password, salt, rounds, bits);
        if (derivationHex.length() > 0) {
            promise.resolve(derivationHex);
            return;
        }
        promise.reject("Key Derivation failed.");
    }

    @ReactMethod
    public void encryptText(
        String encodedText,
        String keyHex,
        String salt,
        String ivHex,
        String hmacKeyHex,
        Promise promise
    ) {
        String encryptedText = Crypto.encryptCBC(
            encodedText,
            keyHex,
            salt,
            ivHex,
            hmacKeyHex
        );
        if (encryptedText == null) {
            promise.reject("Encryption failed.");
            return;
        }
        promise.resolve(encryptedText);
    }

    @ReactMethod
    public void decryptText(
        String encryptedText,
        String keyHex,
        String ivHex,
        String salt,
        String hmacKeyHex,
        String hmacHex,
        Promise promise
    ) {
        String decryptedText = Crypto.decryptCBC(
            encryptedText,
            keyHex,
            ivHex,
            salt,
            hmacKeyHex,
            hmacHex
        );
        if (decryptedText == null) {
            promise.reject("Decryption failed. The archive is possibly tampered with.");
            return;
        }
        promise.resolve(decryptedText);
    }

    @ReactMethod
    public void generateUUIDs(int count, Promise promise) {
        String uuids = Crypto.generateUUIDList(10);
        if (uuids.length() > 0) {
            promise.resolve(uuids);
            return;
        }
        promise.reject("Generating UUIDs failed.");
    }

    @ReactMethod
    public void generateSaltWithLength(int length, Promise promise) {
        String salt = Crypto.generateSalt(length);
        if (salt.length() > 0) {
            promise.resolve(salt);
            return;
        }
        promise.reject("Generating Salt failed.");
    }

    @ReactMethod
    public void generateIV(int length, Promise promise) {
        String ivHex = Crypto.generateRandomBytes(length);
        if (ivHex.length() > 0) {
            promise.resolve(ivHex);
            return;
        }
        promise.reject("Generating IV failed.");
    }

    @Override
    public String getName() {
        return "Crypto";
    }
}
